yarn-project/noir-protocol-circuits/src/crates/rollup-lib/src/base/base_rollup_inputs.nr

```diff
@@ -41,8 +58,373 @@ struct BaseRollupInputs {
+        // 2 unencrypted logs hashes (1 per kernel) -> 4 fields --> 2 sha256 hashes --> 64 bytes
+        let mut calldata_hash_inputs = [0; CALLDATA_HASH_INPUT_SIZE];
+
+        for i in 0..2 {
+            let new_commitments = kernel_data[i].public_inputs.end.new_commitments;
+            let new_nullifiers = kernel_data[i].public_inputs.end.new_nullifiers;
+            let public_data_update_requests = kernel_data[i].public_inputs.end.public_data_update_requests;
+            let newL2ToL1msgs = kernel_data[i].public_inputs.end.new_l2_to_l1_msgs;
+            let encryptedLogsHash = kernel_data[i].public_inputs.end.encrypted_logs_hash;
+            let unencryptedLogsHash = kernel_data[i].public_inputs.end.unencrypted_logs_hash;
+
+            let mut offset = 0;
+
+            for j in 0..MAX_NEW_COMMITMENTS_PER_TX {
+                calldata_hash_inputs[offset + i * MAX_NEW_COMMITMENTS_PER_TX + j] = new_commitments[j];
+            }
+            offset += MAX_NEW_COMMITMENTS_PER_TX * 2;
+
+            for j in 0..MAX_NEW_NULLIFIERS_PER_TX {
+                calldata_hash_inputs[offset + i * MAX_NEW_NULLIFIERS_PER_TX + j] = new_nullifiers[j];
+            }
+            offset += MAX_NEW_NULLIFIERS_PER_TX * 2;
+
+            for j in 0..MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX {
+                calldata_hash_inputs[offset + i * MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX * 2 + j * 2] =
+                    public_data_update_requests[j].leaf_index;
+                calldata_hash_inputs[offset + i * MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX * 2 + j * 2 + 1] =
+                    public_data_update_requests[j].new_value;
+            }
+            offset += MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX * 2 * 2;
+
+            for j in 0..MAX_NEW_L2_TO_L1_MSGS_PER_TX {
+                calldata_hash_inputs[offset + i * MAX_NEW_L2_TO_L1_MSGS_PER_TX + j] = newL2ToL1msgs[j];
+            }
+            offset += MAX_NEW_L2_TO_L1_MSGS_PER_TX * 2;
+
+            let contract_leaf = kernel_data[i].public_inputs.end.new_contracts[0];
+            calldata_hash_inputs[offset + i] = contract_leaf.hash();
+
+            offset += MAX_NEW_CONTRACTS_PER_TX * 2;
+
+            let new_contracts = kernel_data[i].public_inputs.end.new_contracts;
+            calldata_hash_inputs[offset + i * 2] = new_contracts[0].contract_address.to_field();
+            calldata_hash_inputs[offset + i * 2 + 1] = new_contracts[0].portal_contract_address.to_field();
+
+            offset += MAX_NEW_CONTRACTS_PER_TX * 2 * 2;
+
+            for j in 0..NUM_FIELDS_PER_SHA256 {
+                calldata_hash_inputs[offset + i * 2 + j] = encryptedLogsHash[j];
+            }
+
+            offset += NUM_ENCRYPTED_LOGS_HASHES_PER_TX * NUM_FIELDS_PER_SHA256 * 2;
+
+            for j in 0..NUM_FIELDS_PER_SHA256 {
+                calldata_hash_inputs[offset + i * 2 + j] = unencryptedLogsHash[j];
+            }
+        }
+
+        // NOTE!
+        // We deviate from the cpp implementation.
```

Comment:
Noting this: for the small optimisation gain, the cpp code was almost unreadable. It makes sense to do it the non-optimal way first and then optimize